modules/ml/src/main/java/org/apache/ignite/ml/tree/impurity/gini/GiniImpurityMeasureCalculator.java

```diff
@@ -38,54 +41,70 @@
      *
      * @param lbEncoder Label encoder which defines integer value for every label class.
      */
-    public GiniImpurityMeasureCalculator(Map<Double, Integer> lbEncoder) {
+    public GiniImpurityMeasureCalculator(Map<Double, Integer> lbEncoder, boolean useIndex) {
+        super(useIndex);
         this.lbEncoder = lbEncoder;
     }
 
     /** {@inheritDoc} */
     @SuppressWarnings("unchecked")
-    @Override public StepFunction<GiniImpurityMeasure>[] calculate(DecisionTreeData data) {
-        double[][] features = data.getFeatures();
-        double[] labels = data.getLabels();
+    @Override public StepFunction<GiniImpurityMeasure>[] calculate(DecisionTreeData data, TreeFilter filter, int depth) {
+        TreeDataIndex index = null;
+        boolean canCalculate = false;
 
-        if (features.length > 0) {
-            StepFunction<GiniImpurityMeasure>[] res = new StepFunction[features[0].length];
+        if (useIndex) {
+            index = data.createIndexByFilter(depth, filter);
+            canCalculate = index.rowsCount() > 0;
+        }
+        else {
+            data = data.filter(filter);
+            canCalculate = data.getFeatures().length > 0;
+        }
 
-            for (int col = 0; col < res.length; col++) {
-                data.sort(col);
+        if (canCalculate) {
+            int rowsCnt = rowsCount(data, index);
+            int colsCnt = columnsCount(data, index);
 
-                double[] x = new double[features.length + 1];
-                GiniImpurityMeasure[] y = new GiniImpurityMeasure[features.length + 1];
+            StepFunction<GiniImpurityMeasure>[] res = new StepFunction[colsCnt];
 
-                int xPtr = 0, yPtr = 0;
+            for (int col = 0; col < res.length; col++) {
+                if(!useIndex)
+                    data.sort(col);
+
+                ArrayList<Double> x = new ArrayList<>();
```

Comment:
We know size of these arrays, don't we?

Reply:
thanks
I forgot revert it after first iteration of indexes